### PR TITLE
ipn: another controlplane synonym

### LINKED ISSUE #2384: cmd/tailscale: can't change --login-server without --force-reauth
hey @crawshaw, fallout continues:

At 7e7c4c1bbe24f998e76c1d32ef870187829a103d ...

```
bradfitz@tsdev:~/src/tailscale.com$ go run ./cmd/tailscale debug -prefs
{
        "ControlURL": "https://login.tailscale.com",
        "RouteAll": false,
        "AllowSingleHosts": true,
        "ExitNodeID": "",
        "ExitNodeIP": "",
        "ExitNodeAllowLANAccess": false,
        "CorpDNS": true,
        "WantRunning": true,
        "LoggedOut": false,
        "ShieldsUp": false,
        "AdvertiseTags": null,
        "Hostname": "",
        "OSVersion": "",
        "DeviceModel": "",
        "NotepadURLs": false,
        "AdvertiseRoutes": [],
        "NoSNAT": false,
        "NetfilterMode": 2,
        "OperatorUser": "bradfitz",
        "Config": {
                "PrivateMachineKey": "privkey:0000000000000000000000000000000000000000000000000000000000000000",
                "PrivateNodeKey": "privkey:0000000000000000000000000000000000000000000000000000000000000000",
                "OldPrivateNodeKey": "privkey:0000000000000000000000000000000000000000000000000000000000000000",
                "Provider": "google",
                "LoginName": "bradfitz@tailscale.com"
        }
}

bradfitz@tsdev:~/src/tailscale.com$ go run ./cmd/tailscale up --accept-dns=false 
can't change --login-server without --force-reauth
exit status 1

bradfitz@tsdev:~/src/tailscale.com$ go run ./cmd/tailscale up --accept-dns=false -h 2>&1 | grep -A 1 login-server
  --login-server string
        base URL of control server (default https://controlplane.tailscale.com)
```

### LINKED ISSUE #2386: CSRF error after cmd/tailscale up sends me to controlplane.tailscale.com to authorize machines
```
$ tailscale up ...
...
To authorize your machine, visit (as admin):
 
        https://controlplane.tailscale.com/admin/machines
```

Go through Google oauth flow, and then:

Screen Shot 2021-07-10 at 9 39 56 PM

### PR DESCRIPTION
**Pull request by crawshaw**
This one doesn't bother me so much, as long term we want a synonym here.

Fixes #2384
Fixes #2386

Signed-off-by: David Crawshaw <crawshaw@tailscale.com>

### CODE REVIEW

File: cmd/tailscale/cli/up.go [hunk 1]
```diff
@@ -307,7 +307,7 @@ func runUp(ctx context.Context, args []string) error {
 		}
 	}
 
-	controlURLChanged := curPrefs.ControlURL != prefs.ControlURL
+	controlURLChanged := curPrefs.ControlURL != prefs.ControlURL || (ipn.IsLoginServerSynonym(curPrefs.ControlURL) && ipn.IsLoginServerSynonym(prefs.ControlURL))
```
Comment (bradfitz) [inline cmd/tailscale/cli/up.go:310]: is there a missing `!` in here? this doesn't read right to me.

Also: no tests. Can you add tests to convince me this is correct? It's already table-driven so should be easy to add some more table entries.

I assume this is the part of the fix that handles the "up" by itself shouldn't do anything?

Comment (crawshaw) [reply cmd/tailscale/cli/up.go:310]: The current tests don't seem to cover this code. But I suppose they should, I'll figure out a way to do that.

File: cmd/tailscale/cli/up.go [hunk 2]
```diff
@@ -259,7 +259,8 @@ func updatePrefs(prefs, curPrefs *ipn.Prefs, env upCheckEnv) (simpleUp bool, jus
 		}
 	}
 
-	controlURLChanged := curPrefs.ControlURL != prefs.ControlURL
+	controlURLChanged := curPrefs.ControlURL != prefs.ControlURL &&
+		!(ipn.IsLoginServerSynonym(curPrefs.ControlURL) && ipn.IsLoginServerSynonym(prefs.ControlURL))
```
Comment (bradfitz) [inline cmd/tailscale/cli/up.go:263]: there's the `!` :)

File: ipn/prefs.go [hunk 3]
```diff
@@ -30,6 +30,10 @@ import (
 // The default control plane is the hosted version run by Tailscale.com.
 const DefaultControlURL = "https://controlplane.tailscale.com"
 
+func IsLoginServerSynonym(val interface{}) bool {
```
Comment (bradfitz) [inline ipn/prefs.go:33]: docs

File: ipn/prefs.go [hunk 4]
```diff
@@ -405,6 +409,15 @@ func (p *Prefs) ControlURLOrDefault() string {
 	return DefaultControlURL
 }
 
+func (p *Prefs) AdminPageURL() string {
```
Comment (bradfitz) [inline ipn/prefs.go:412]: docs

**Review by bradfitz — APPROVED** [review]

File: cmd/tailscale/cli/up.go (hunk 2 above)
Comment (crawshaw) [reply cmd/tailscale/cli/up.go:263]: sigh, tests

**Pull request merged**